engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java

```diff
@@ -98,4 +100,56 @@ public void ascribedWithAParameter() throws Exception {
     var some = module.invokeMember("eval_expression", "Maybe.Some 10");
     assertEquals("Can read ten", 10, some.getMember("unwrap").asInt());
   }
+
+  @Test
+  public void binaryWithZero() throws Exception {
+    Value module = exampleWithBinary();
+    var ok = module.invokeMember("eval_expression", "Bin.Zero Zero");
+    assertEquals("binary.Bin", ok.getMetaObject().getMetaQualifiedName());
+    try {
+      var v = module.invokeMember("eval_expression", "Bin.Zero 'hi'");
+      fail("Expecting an error, not " + v);
+    } catch (PolyglotException ex) {
+      assertEquals("Type_Error.Error", ex.getMessage());
+    }
+    try {
+      var v = module.invokeMember("eval_expression", "Bin.Zero One");
+      fail("Expecting an error, not " + v);
+    } catch (PolyglotException ex) {
+      assertEquals("Type_Error.Error", ex.getMessage());
+    }
+  }
+
+  private Value exampleWithBinary() throws URISyntaxException {
+    var uri = new URI("memory://binary.enso");
+    var src = Source.newBuilder("enso", """
+    type Zero
+    type One
+
+    type Bin
+        Zero (v:Zero)
+        One (v:One)
+        Either v:(Zero | One)
+    """, uri.getHost()).uri(uri).buildLiteral();
+    return ctx.eval(src);
+  }
+
+  @Test
+  public void binaryWithOne() throws Exception {
+    Value module = exampleWithBinary();
+    var ok = module.invokeMember("eval_expression", "Bin.One One");
+    assertEquals("binary.Bin", ok.getMetaObject().getMetaQualifiedName());
+    try {
+      var v = module.invokeMember("eval_expression", "Bin.One 10");
+      fail("Expecting an error, not " + v);
+    } catch (PolyglotException ex) {
+      assertEquals("Type_Error.Error", ex.getMessage());
+    }
+    try {
+      var v = module.invokeMember("eval_expression", "Bin.One Zero");
+      fail("Expecting an error, not " + v);
+    } catch (PolyglotException ex) {
+      assertEquals("Type_Error.Error", ex.getMessage());
+    }
+  }
```

Comment:
Btw. do you think it could be useful to also allow syntax like:
```
foo (v : Vector Integer) =
    x : Integer = v.first
    y : Integer = v.first
    x + y
```
i.e. an ascription at local variable definition could also verify the type.

Not for this PR, since I checked and currently:
```
> (x : Integer) = 42
Evaluation failed with: Syntax error: Syntax is not supported yet: qualifiedNameSegment.
```
so we would need some parser improvements first.

But I think it could be a good next step to allow us to easily 'assert' types, not only in public signatures but also in the implementation.

Reply:
Reported as #6848

Reply:
> Reported as #6848

Thanks, that is one part of the question.

Another is:

> From the PR description, this also implements typechecking for ascriptions in methods, right? But I don't see any tests showing it.
> 
> Unless I'm missing something, can we get a test case for functions?

I looked through the PR and cannot find such test cases. I still might have missed something. But if not, can we get a test case?

Currently, it looks as if the PR only adds checking to _constructors_ - which is **great** but it also says it adds checking to regular methods - but I cannot see any test verifying that.